content/en/blog/2023-01-21-prototyping-eda-with-watermill.md

```diff
@@ -0,0 +1,173 @@
+
+<!--more-->
+
+Event-driven architecture is a powerful framework for building modern applications. Unlike conventional applications, which follow a request-response pattern, EDA enables the decoupling of software components, allowing them to asynchronously pass data between each other. EDA often helps reduce latency and increase throughput and is well suited for modern, distributed system architectures. It also opens up opportunities for organizations to build more responsive applications and more quickly integrate new features.
+
+## Eventing is Hard, Watermill Makes it a Little Easier
+
+While there are many ways to build event-driven applications (check out [this post](https://rotational.io/blog/eventing-platforms/) to learn more about the types of tools available), building event-driven applications is still pretty hard. It requires advanced software engineering skills and DevOps skills to get projects up and running.  On top of that, there is a lack of standardization among tools, so it is difficult to make changes to tooling after the fact (e.g. once a team has a better understanding of the throughput/latency needs of their app).
+
+This is where [Watermill](https://watermill.io) comes in. Watermill is an open source Golang package that enables developers unfamiliar with EDA to get started building event-driven applications. Furthermore, it provides a standardized API that allows developers to experiment with different EDA tools. It is very simple to start building applications in one eventing tool, and with just a few changes, switch to another. The goal of Watermill, as the creators have stated, is "to make communication with messages as easy to use as HTTP routers." They go on to say:
+
+> Just like you don’t need to know the whole TCP stack to create a HTTP REST server, you shouldn’t need to study all of this knowledge to start with building message-driven applications.
+
+## The Basics of Eventing
+
+At the core of EDA is the **event** (or **message**), for example, a customer adding an item to a shopping cart or a smart light being turned on. In the EDA framework, a **producer** (or **publisher**) writes these events to a **topic**, while a **consumer** (or **subscriber**) listens to the topic stream, consuming and processing messages as they arrive.
+
+There are several advantages to this framework. First, the producer and consumer applications can be built independently of each other and the processing can be asynchronous. There is no need for blocking code execution as there is in a traditional request/response framework, where the client has to block until it gets a response from the server. Second, the producer application can simply send data in its raw form, and it is up to the consumer application to do the data transformations that is suited for its use case. This opens up the possibilities of having multiple consumers listening on the same topic to serve different business functions.
+
+## Getting Started with Watermill
```

Comment:
The only thing I was thinking we should add here is a little note that the subsequent code snippets assume that the reader has installed watermill, ensign, and watermill-ensign.